lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -329,23 +329,37 @@ LogicalResult LowerLayersPass::runOnModuleBody(FModuleOp moduleOp,
   StringRef circuitName = circuitOp.getName();
   hw::InnerSymbolNamespace ns(moduleOp);
 
-  // A cache of nodes we've created for values which are captured, but cannot
-  // have a symbol attached to them.  This is done to avoid creating the same
-  // node multiple times.
-  DenseMap<Value, NodeOp> nodeCache;
-
-  // Get or create a node within this module.  This is used when we need to
-  // create an XMRDerefOp to an expression.
-  auto getOrCreateNodeOp = [&](Value operand, ImplicitLocOpBuilder &builder,
-                               StringRef nameHint) -> NodeOp {
-    auto it = nodeCache.find(operand);
-    if (it != nodeCache.end())
+  // A cache of values to nameable ops that can be used
+  DenseMap<Value, Operation *> nameableCache;
+
+  // Get or create a node op for a value captured by a layer block.
+  auto getOrCreateNodeOp = [&](Value operand,
+                               ImplicitLocOpBuilder &builder) -> Operation * {
+    // Use the cache hit.
+    auto it = nameableCache.find(operand);
```

Comment:
Consider a variant that does a since lookup, and update the value if cache doesn't have this operand already.

Reply:
This is interesting... For some reason this pattern never occurred to me to do. Just for clarity, you're talking about something like:

``` c++
auto *op = foo.lookup(...);
if (op)
  return op;

/* body */

return op = builder.create<T>(...);
```

I normally don't consider this as I didn't think there was a default constructor here, but I guess the default would be a `nullptr` `Operation *` so this whole thing slots together nicely.

Do I understand you correctly?

Originally, I thought you were advising a hypothetical `lookupOrElse(key, function_ref)` which I didn't think existed (though I'm used to seeing in other languages).

Reply:
Changed in: https://github.com/llvm/circt/pull/8466/commits/1028e647fdc95b8dede529e503cfa190e35f4526